### ISSUE TITLE
afterExecuteQuery ran before query end

### ISSUE THREAD

**Issue by deblockt** [r2dbc/r2dbc-proxy#94]
#### Versions

- Driver: r2dbc-postgres 0.8.6 poxy 0.8.3 spi 0.8.3
- Database: postgres
- Java: java 11
- OS: max

#### Current Behavior

the afterExecuteQuery is called before end of the query execution mono, so when we use it to report execution time, the result is bad

```
2021-06-04 12:21:12.202 START THE QUERY
2021-06-04 12:21:12.202 Executing SQL statement [select data from myTable where id = :id LIMIT 1]
2021-06-04 12:21:12.203 Expanding SQL statement [select data from myTable where id = :id LIMIT 1] to [select data from myTable where id = $1 LIMIT 1]
2021-06-04 12:21:12.203 beforeExecuteQuery > start the query select data from myTable where id = $1 LIMIT 1
2021-06-04 12:21:12.246 afterExecuteQuery > end the query select data from myTable where id = $1 LIMIT 1
2021-06-04 12:21:12.687 DATA LOADED
```
On this logs we can see the the afterExecuteQuery is called 400ms before the end of the mono.

#### Steps to reproduce

This code is used to reproduce the issue: 
``` java
final var sql = "select data from myTable where id = :id LIMIT 1";

return MonoUtil.infoLog("START THE QUERY")
    .then(
      this.databaseClient.execute(sql)
    .bind("id", myObject.toString())
    .map(row -> row.get("data", String.class))
    .one())
    .doOnNext(paymentExecution -> log.info("DATA LOADED"));
```

I'm not sur that the issue is on this project or on the postgres driver. 

The query is slow because the table is bug > 3Go and an index is missing. 

#### Expected behavior/code

the method afterExecuteQuery should be called when the query is ended and the result fetched.

**Comment by deblockt** [comment]
I tried to use the `ProxyMethodExecutionListener` instead of the old `LifeCycleListener` but the result is same.
I tried with this code: 
```

@Slf4j
public class TracingExecutionListener implements ProxyMethodExecutionListener {

    @Override
    public void beforeQuery(QueryExecutionInfo execInfo) {
        final var queries = execInfo.getQueries().stream()
            .map(QueryInfo::getQuery)
            .collect(joining(", "));
        log.info("TracingExecutionListener beforeQuery > start the query {}", queries);
    }

    @Override
    public void eachQueryResult(QueryExecutionInfo execInfo) {
        final var queries = execInfo.getQueries().stream()
            .map(QueryInfo::getQuery)
            .collect(joining(", "));
        log.info("TracingExecutionListener eachQueryResult > a result on the query {} {}", queries, execInfo.getExecuteDuration().toMillis());
    }

    @Override
    public void afterQuery(QueryExecutionInfo execInfo) {
        final var queries = execInfo.getQueries().stream()
        .map(QueryInfo::getQuery)
        .collect(joining(", "));
        log.info("TracingExecutionListener afterQuery > end the query {} {}ms", queries, execInfo.getExecuteDuration().toMillis());
    }
}

```

And I have a strange workflow on logs
```
2021-06-04 16:07:16.679 START THE QUERY
2021-06-04 12:21:12.202 Executing SQL statement [select data from myTable where id = :id LIMIT 1]
2021-06-04 12:21:12.203 Expanding SQL statement [select data from myTable where id = :id LIMIT 1] to [select data from myTable where id = $1 LIMIT 1]
2021-06-04 16:07:16.711 TracingExecutionListener beforeQuery > start the select data from myTable where id = $1 LIMIT 1
2021-06-04 16:07:16.727 TracingExecutionListener afterQuery > end the select data from myTable where id = $1 LIMIT 1 15ms
2021-06-04 16:07:17.336 TracingExecutionListener eachQueryResult > a result on the select data from myTable where id = $1 LIMIT 1 15
```

the `afterQuery` is called before the `eachQueryResult` and the request duration seem bad

**Comment by ttddyy** [comment]
Hi @deblockt 
Thanks for the report, but I couldn't reproduce the issue.
Can you provide a minimal repro project?

Also, can you confirm `execInfo.isSuccess()` in `afterQuery()`?

**Comment by deblockt** [comment]
Hi, I can't reproduce on a minimal project (without spring). In debug mode I can see that the issue come from a canceled mono.

**Comment by deblockt** [comment]
I'm creating a reproducer projet. 
I can see that the issue happen only when using a binded var. I try using spring. 
This code works: 
```
        return Mono.fromRunnable(() -> log.info("before executing query mono"))
            .then(
                databaseClient.execute("select data from payment_execution where data->>'paymentScheduleId' = 'b8b4304f-9db7-4c8c-b690-4e6338a4de83' FOR UPDATE LIMIT 1")
                    .map(row -> row.get("data", String.class))
                    .one()
            ).doOnNext($ -> log.info("after execute the query mono"));
```

This code produce the issue:
```
        return Mono.fromRunnable(() -> log.info("before executing query mono"))
            .then(
                databaseClient.execute("select data from payment_execution where data->>'paymentScheduleId' = :id FOR UPDATE LIMIT 1")
                    .bind("id", "b8b4304f-9db7-4c8c-b690-4e6338a4de83")
                    .map(row -> row.get("data", String.class))
                    .one()
            ).doOnNext($ -> log.info("after execute the query mono"));
```

**Comment by mp911de** [comment]
Note that R2DBC doesn't specify the time of the actual query execution. In most cases, `Statement.execute()` returns a `Publisher<Result>` that represents a promise for statement execution. It performs the actual query upon subscription.

The reason is that creating a `Result` doesn't necessarily require database interaction as `Result` semantically translates to `Publisher<Row>` (or `Publisher<UpdateCount>`). While at first sight, it isn't obvious why this indirection has to happen, consider an arrangement with multiple bindings (`Statement.add()`):

```java
connection.createStatement("SELECT … WHERE field = $1")
    .bind("$1", "first").add()
    .bind("$2", "second").execute();
```

What should be the outcome? When should be `SELECT … WHERE field = 'second'` run? On `.execute()` or after `SELECT … WHERE field = 'first'` was fully consumed (potentially by using cursors)? 

Depending on how a database identifies cursors and how many open cursors there can be, the answer might vary. This isn't just a theoretical problem. In an arrangement using auto-commit, the outcome of the second query can vary depending on the time window that elapses between the first and second query.

**Comment by deblockt** [comment]
I created a sample project to reproduce the issue: https://github.com/deblockt/r2dbc-issue-reproducer

**Comment by deblockt** [comment]
I'm looking the r2dbc-proxy code and I understand the answer of @mp911de . 
Actually, to know if the query is "ended" we "observe" the `complete` of the `Mono<Result>` but the `Mono<Result>` is complete before the real query execution. 
If we want to know when the query is ended, we should look all `Mono` of `Result.map` or `Result.getRowsUpdated` methods.

**Comment by mp911de** [comment]
> If we want to know when the query is ended, we should look all Mono of Result.map or Result.getRowsUpdated methods.

Yes, exactly. Note that drivers can make use of cursors. Depending on what you define as "query has ended", you can measure the time until the first result or `onComplete(…)` of `Result.map` or `Result.getRowsUpdate`

**Comment by ttddyy** [comment]
@deblockt Thanks for the update and @mp911de thanks for the detailed explanation

Based on Mark's explanation, yes I agree current timing to call `afterQuery` callback needs to move to the completion (and cancel/error) of `Result.map` and `Result.getRowsUpdated` publisher.

But now it makes me think what about if user doesn't consume the `Result`.

For example:
```java
connection.createStatement("select pg_sleep(1)").execute().then(...)
```

In this case, it won't call `Result.[map|getRowsUpdate]`.

I think I need to think through the behavior of the `afterQuery` callback and the definition itself.

Possible change may be `afterQuery` to be called at the end of consuming the `Statement.execute()` publisher, and may need to introduce `beforeResult` and `afterResult` that ties to the consumption of `Result.[map|getRowsUpdate]` publisher.

Currently, I'm pretty much overloaded by my day-to-day work.
So, allow me to take additional time to start working on this.

**Comment by deblockt** [comment]
I think that call `.map` or `.getRowUpdate` is mandatory on r2dbc. 
I just perform a test and see that: 
   - `connection.createStatement("select pg_sleep(1)").execute().then(...)` work the query is executed 
   - `connection.createStatement("select pg_sleep($1)").bind(0, 3).execute().then(...)` not work the query is not executed

This is because the query returned by `createStatement` contains de `Result` and as marc said

> The reason is that creating a Result doesn't necessarily require database interaction as Result semantically translates to Publisher<Row> (or Publisher<UpdateCount>). While at first sight, it isn't obvious why this indirection has to happen, consider an arrangement with multiple bindings (Statement.add()):

So I think that the example `connection.createStatement("select pg_sleep(1)").execute().then(...)` is "bad" and should not be used without map the `Mono<Result>`

Looking on the Spring code, we can see that the `.then` do that `this.fetch().rowsUpdated().then()` so using spring the `rowUpdated` is called on each cases. 

I think that since the query is potentially no executed on the case `.execute().then(...)` the call of `afterQuery` can be a bad idea. maybe that we should have another method like `afterQueryInit` or something like that.

@mp911de what do you think about the `.execute().then(...)` can we distinguish the case where the query is "executed" without call the `.map` and the case where the query is not "executed"


@ttddyy I will try to provide a PR on my side to provide a beginning of solution for this issue

**Comment by deblockt** [comment]
I created a draft pull request https://github.com/r2dbc/r2dbc-proxy/pull/95 with a first approach to solve the issue

**Comment by mp911de** [comment]
> I think that call .map or .getRowUpdate is mandatory on r2dbc.

Exactly. Without that, the result isn't consumed and because of no materialization of the result, the command either doesn't get executed or data remains on the transport channel (no demand and no one is consuming data).

`.execute().then(...)` is a Reactor feature. It is unsafe because the result doesn't get consumed.

**Comment by ttddyy** [comment]
@deblockt @mp911de Thanks for the discussion and draft PR.

I have created a failing test simulating the situation here and pushed to my [gh-94_0.8.x_after-query](https://github.com/ttddyy/r2dbc-proxy/tree/gh-94_0.8.x_after-query) branch.

About the implementation, I am thinking rather moving the `[before|after]-query` callbacks to the subscriber of `Result#[map|getRowUpdated]` publisher, compared to the current implementation in the subscriber of `Statement#execute`.

As discussed above, the entry point for triggering the query execution is actually the subscription of the publisher from `Result#[map|getRowUpdated]`. So, I think this is more natural to start the before-query callback at the subscription of `Result#[map|getRowUpdated]` and end with after-query callback at the completion/cancel/error of the `Result#[map|getRowUpdated]` publisher.

**Comment by deblockt** [comment]
If we move the `[before|after]-query` callback to the `Result#[map|getRowUpdated]` publisher we need 1 `MutableQueryExecutionInfo` by generated `Result` (I think that is needed to avoid conflict if two `Result.map` are ran in parallel). But I see that `MutableQueryExecutionInfo` is generated on `Statement.execute` call to get informations about executed query. 

If we use this query: 
``` java
Flux.from(connection.createStatement("select pg_sleep($1), '1' as data")
                    .bind(0, 3)
                    .add()
                    .bind(0, 1)
                    .execute())
                    .flatMap(result -> {
                        return result.map((row, it) -> row.get("data", String.class));
                    });
```
With your approach give something like that: 
```
=> beforeQuery (select pg_sleep($1), '1' as data, binding: [0, 3], [0, 1]) 
=> eachQueryResult (select pg_sleep($1), '1' as data, binding: [0, 3], [0, 1]) 
=> afterQuery (select pg_sleep($1), '1' as data, binding: [0, 3], [0, 1]) 
=> beforeQuery (select pg_sleep($1), '1' as data, binding: [0, 3], [0, 1]) 
=> eachQueryResult (select pg_sleep($1), '1' as data, binding: [0, 3], [0, 1]) 
=> afterQuery (select pg_sleep($1), '1' as data, binding: [0, 3], [0, 1]) 
```

In the current approach implemented on my PR, the result with be: 
```
=> beforeQuery (select pg_sleep($1), '1' as data, binding: [0, 3], [0, 1]) 
=> eachQueryResult (select pg_sleep($1), '1' as data, binding: [0, 3], [0, 1]) 
=> eachQueryResult (select pg_sleep($1), '1' as data, binding: [0, 3], [0, 1]) 
=> afterQuery (select pg_sleep($1), '1' as data, binding: [0, 3], [0, 1]) 
```

For this approach the code is more complex (the after query can be called on by `Result.map` or by the `Mono<Result>.onComplete`. 

I don't know what is the best solution?

**Comment by ttddyy** [comment]
@deblockt Thanks for the details. It helps me a lot.

I think it will be similar to what your draft PR does.
Instead of adding methods to `ExecutionInfo`, probably introduce a reference object that holds counter(s) to keep track of how many results are produced and processed. The subscriber of `Statement#execute` and `Result#[map|getRowsUpdated]` will share this information and determine when is the actual end of the consumption of all results.

**Comment by deblockt** [comment]
🎉 
@ttddyy do you know when the release will be done?

**Comment by ttddyy** [comment]
@deblockt I have released `0.8.7.RELEASE`
